### PR TITLE
os_unfair_lock

### PR DESCRIPTION
**Pull request by moto0000**
Few days ago I found a very interesting way to make locekd value:
https://github.com/SvenTiigi/CodableStoreKit/blob/master/Sources/Extensions/Locked.swift

Benchmark:
https://gist.github.com/steipete/36350a8a60693d440954b95ea6cbbafc

Article:
https://www.mikeash.com/pyblog/friday-qa-2017-10-27-locks-thread-safety-and-swift-2017-edition.html

### CODE REVIEW

**Comment by kean** [comment]
Thanks, @moto0000. I'm considering using spin locks. I think it can be a major win in this case. Thanks for putting this together.

File: Sources/Future.swift [hunk 1]
```diff
@@ -366,49 +368,43 @@ extension Future {
...
-            lock.unlock()
 
             inlinedHandler?(result)
             handlers?.forEach { $0(result) }
         }
 
         func observe(completion: @escaping (Result) -> Void) {
-            lock.lock()
-            guard let result = self.memoizedResult else {
+            guard let result = self.memoizedResult.value else {
```
Comment (kean) [inline Sources/Future.swift]: AFAIR this lock is used not only to synchronize access to `memoizedResult`, but also to `handlers` and `inlineHandler`.

File: Sources/Future.swift [hunk 2]
```diff
@@ -341,13 +341,15 @@ extension Future {
 
     /// A promise to provide a result later.
     public final class Promise: CustomDebugStringConvertible {
-        private var memoizedResult: Result? // nil when pending
+        private var memoizedResult: Locked<Result?> // nil when pending
         private var inlinedHandler: ((Result) -> Void)?
         private var handlers: [(Result) -> Void]?
         private let lock = NSLock()
```
Comment (kean) [inline Sources/Future.swift]: This lock should also be removed in favor of spin lock. This is going to be major win! Instantiating `NSLock` is relatively expensive.

**Review by kean — CHANGES_REQUESTED** [review]
I'm in favor of merging this. This was the last thing that I wanted to optimize for a while. I think after this change, FutureX might be *the fastest* future implementation out there. Thanks again, @moto0000, for putting this together.

**Comment by moto0000** [comment]
@kean I make few changes

**Review by kean — APPROVED** [review]

**Comment by kean** [comment]
Hey, @moto0000, I ran the performance tests included in the repo and I haven't noticed any performance improvements. I tried using Nuke repo, I also didn't see any performance improvements. These results in [RxSwift](https://github.com/ReactiveX/RxSwift/pull/2042) repo also suggest that there are no noticeable performance improvements.

I think Steinberger's results are no longer relevant. I know for a fact that `NSLock` used to be slow but somewhere in 2016 Apple rewrote it and it now has performance comparable to spin locks in many cases.

I think I will revert these changes and stick with `NSLock` just to be safe, since it's a higher level primitive. There are less chances to run into issue with `NSLock` than with `os_unfair_lock_s`.

**Pull request merged**